v4: Add interaction utils

mdo opened the pull request:
This might be overstepping things for our next v4 release, but I noticed how highly requested the `.cursor-pointer` utility was in #23709 and couldn't help but stub out a PR to address it. I've taken the v5 approach of an interactions page and included this new util, and the `user-select` one from v5. Decided not to include `pointer-events` utils because of browser compatibility.

I'd really like to ship small enhancements to v4 for folks with highly requested features like this, especially if it helps bridge the gap between v4 and v5. My biggest question is how we version/release this though.

Do we squeeze this into a v4.4.2 despite of it being a new feature and thus requiring a semver minor bump? Or do we upgrade v4.4.2 to v4.5?

XhmikosR commented:
Squashed the patches.

TBH I'd prefer it if we didn't make it a new minor release. On the other hand, semver...

Let's wait for the others to chime in.

MartijnCuppens commented:
> I'd prefer it if we didn't make it a new minor release

I agree

mdo commented:
Revisiting this, if @twbs/css-review is into it, I'm inclined to sneak this into v4.4.2 to help close the gap between v4 and v5. Really think this will be some solid good will for our users :).

XhmikosR commented:
We never followed semver 100% and this, at least, is not a breaking change, so I think we can add it.

MartijnCuppens commented:
I'm gonna stick with what I said in https://github.com/twbs/bootstrap/pull/30563#issuecomment-613005377. I don't want people to abuse this property.

What I am willing to consider is adding
```
[role="button"] {
  cursor: pointer;
}
```
to reboot since this will be required anyway.

XhmikosR commented:
That's a valid concern of course, but on the other hand, one can buy a gun legally (in some countries); if they kill someone the store cannot be held liable for any murders. 😇

But I definitely see your point. Maybe if we added a warning callout it'd be better?

ffoodd commented:
I'd be on both sides on this one: 
- I agree with @MartijnCuppens about `cursor`s and don't think it should be a utility at all. I already know tons of people who'd just add a `div` with `.btn` things and `.cursor-pointer`—and it'll probably be bad. `[role="button"]` seems more accurate.
- `user-select` on the other hand seems handy, I don't see any weird use cases (but I may be a bit naive on this).

mdo commented:
Updated to reflect the team's feedback! Thanks y'all.

ffoodd commented:
Shouldn't `[role="button"]` sit in reboot, as suggested by @MartijnCuppens? I think it shouldn't be considered as a utility since it needs accessibility and scripting considerations to ne used the right way.

Maybe even provide a sentence with explanation or link?

ffoodd commented on scss/utilities/_interactions.scss:
```diff
@@ -0,0 +1,9 @@
+  .user-select-#{$value} { user-select: $value !important; }
+}
+
+[role="button"] {
+  cursor: pointer;
+}
```
I think this should sit in `_reboot.scss`, and not being encouraged.

ffoodd commented on site/docs/4.4/utilities/interactions.md:
```diff
@@ -0,0 +1,30 @@
+
+Customize the available classes by changing the `$user-selects` Sass list in `_variables.scss`.
+
+## Cursors
+
+Change the cursor from a default value to `pointer` with our built-in attribute utility, `role="button"`. Add this attribute to an element to convey semantics to your visitors and add an explicit pointer cursor to help indicate elements are interactive.
+
+{% capture example %}
+<span role="button">Non-button element button</span>
+{% endcapture %}
+{% include example.html content=example %}
+
+Customize the available classes by changing the `$cursors` Sass list in `_variables.scss`.
```
Since this part only mentions `[role="button"]`, I'd be in favor of dropping it totally.

Using this `role` has multiple prerequisite, like ensuring the interaction layer respects the `button` pattern: activation with <kbd>Space</kbd> and <kbd>Enter</kbd>, at least. This requires JavaScript and has very strong accessibility concerns: I wouldn't recommend to use this attribute to change the `cursor`.

Adding this `cursor` thing to our reboot seems good to prevent people from adding it manually… But this is more a bad practice than anything, and a `button` element should be used instead.We cannot recommend this.

BTW the last sentence is outdated anyway, isn't it?

mdo committed:
Add new interaction utilities and role attr to Reboot

- Adds .user-select-* utils from v5
- Adds button role attribute util, one of the top requested features in our issues for adding pointer cursors
- Adds new docs page to demonstrate both
- Includes Sass list for customizing user select

mdo commented:
Think I got the feedback applied @MartijnCuppens and @ffoodd. Let me know with any suggested changes or review comments if there's something I can clarify more in the docs <3.

ffoodd reviewed (approved):
LGTM. We'll probably add some warning about the button role usage, but I think it's more @patrickhlauke's business here.

MartijnCuppens reviewed (approved):
:shipit:

MartijnCuppens committed:
Merge branch 'v4-dev' into v4-interaction-utils

XhmikosR commented:
FYI the utilities/interactions.scss wasn't included in utilities.scss. I fixed this in my v4-dev PR #30768.
